halo2/src/mock_prover.rs

```diff
@@ -0,0 +1,139 @@
+use std::{fs::File, path::Path};
+
+use number::read_polys_file;
+use pil_analyzer::Analyzed;
+use polyexen::plaf::PlafDisplayBaseTOML;
+
+use super::circuit_builder::analyzed_to_circuit;
+use halo2_proofs::{dev::MockProver, halo2curves::bn256::Fr};
+use number::{BigInt, Bn254Field, FieldElement};
+
+pub fn mock_prove(file: &Path, dir: &Path) {
+    let analyzed: Analyzed<Bn254Field> = pil_analyzer::analyze(file);
+
+    assert_eq!(
+        polyexen::expr::get_field_p::<Fr>(),
+        Bn254Field::modulus().to_arbitrary_integer(),
+        "powdr modulus doesn't match halo2 modulus"
+    );
+
+    let fixed_columns: Vec<&str> = analyzed
+        .constant_polys_in_source_order()
+        .iter()
+        .map(|(poly, _)| poly.absolute_name.as_str())
+        .collect();
+
+    let witness_columns: Vec<&str> = analyzed
+        .committed_polys_in_source_order()
+        .iter()
+        .map(|(poly, _)| poly.absolute_name.as_str())
+        .collect();
+
+    let (fixed, fixed_degree) = read_polys_file(
+        &mut File::open(dir.join("constants").with_extension("bin")).unwrap(),
+        &fixed_columns,
+    );
+    let (witness, witness_degree) = read_polys_file(
+        &mut File::open(dir.join("commits").with_extension("bin")).unwrap(),
+        &witness_columns,
+    );
+
+    assert_eq!(fixed_degree, witness_degree);
+
+    mock_prove_ast(&analyzed, fixed, witness)
+}
+
+fn mock_prove_ast<T: FieldElement>(
+    pil: &Analyzed<T>,
+    fixed: Vec<(&str, Vec<T>)>,
+    witness: Vec<(&str, Vec<T>)>,
+) {
+    let circuit = analyzed_to_circuit(pil, fixed, witness);
+
+    // double the row count in order to make space for the cells introduced by the backend
+    // TODO: use a precise count of the extra rows needed to avoid using so many rows
+
+    let circuit_row_count_log = usize::BITS - circuit.plaf.info.num_rows.leading_zeros();
+
+    let expanded_row_count_log = circuit_row_count_log + 1;
+
+    log::debug!("{}", PlafDisplayBaseTOML(&circuit.plaf));
+
+    let inputs = vec![];
+
+    let mock_prover = MockProver::<Fr>::run(expanded_row_count_log, &circuit, inputs).unwrap();
+    mock_prover.assert_satisfied();
+}
+
+#[cfg(test)]
+mod test {
+    use std::fs;
+
+    use super::*;
+
+    fn mock_prove_asm(file_name: &str, inputs: &[Bn254Field]) {
+        // read and compile PIL.
+
+        let contents = fs::read_to_string(file_name).unwrap();
+        let pil = pilgen::compile::<Bn254Field>(Some(file_name), &contents).unwrap_or_else(|err| {
+            eprintln!("Error parsing .asm file:");
+            err.output_to_stderr();
+            panic!();
+        });
+
+        let query_callback = |query: &str| -> Option<Bn254Field> {
```

Comment:
It would be nice to reduce code duplication here, especially since I'm changing the query format from time to time...

Reply:
I'm also fine with moving this to witgen and have one "default" query callback that takes a vector of numbers, but also the option for the user to change it if they do custom pil.

Reply:
Haha yes an outdated query format made me waste an hour when working on this PR. I think having a default one makes sense. I opened https://github.com/powdr-org/powdr/issues/302 for it.